### PR TITLE
Adds structure and algorithm

### PR DESCRIPTION
**Pull request by JosephSalisbury**
Towards https://github.com/giantswarm/giantswarm/issues/3183

Basic cobra setup, notes for scheduling algorithm.

```
$ ./oncall-scheduler schedule --help
Ensure oncallers are scheduled

Usage:
  oncall-scheduler schedule [flags]

Flags:
  -h, --help                    help for schedule
      --horizon duration        How far ahead to ensure oncallers are scheduled (default 168h0m0s)
      --past-horizon duration   How far back to look for scheduling (default 672h0m0s)
      --schedule-name string    OpsGenie schedule to ensure (default "ops_schedule")
      --team-name string        OpsGenie team to draw oncallers from (default "ops_team")
```

### CODE REVIEW

**Review by xh3b4sd — COMMENTED** [review]
It would be really nice if we could go with the https://github.com/giantswarm/devctl structure. There is this runner concept. IMO we should align our command line tools accordingly. These things get out of hand.

**Comment by JosephSalisbury** [comment]
@xh3b4sd || @kopiczko ptalaaa

File: cmd/root.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+	if config.Stderr == nil {
+		config.Stderr = os.Stderr
+	}
+	if config.Stdout == nil {
+		config.Stdout = os.Stdout
+	}
+
+	if config.GitCommit == "" {
```
Comment (xh3b4sd) [inline cmd/root.go]: ~~We should may make use of the `project` package we started to introduce with VOO. See also what Theo announced: https://gigantic.slack.com/archives/C04TGHDEF/p1560515977009100.~~

---

Hohoho I saw you are a good boy and added the `project`. Nice! Then we do not need to configure these things but only use the package interface to get information where they are required.

File: main.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+			Source:    project.Source(),
+		}
+
+		rootCommand, err = cmd.New(c)
+	}
+
+	err = rootCommand.Execute()
+	if err != nil {
+		logger.LogCtx(ctx, "level", "error", "message", "failed to execute command", "stack", microerror.Stack(err))
+		os.Exit(1)
```
Comment (xh3b4sd) [inline main.go]: The reason for `mainE` is that we can return an error and do logging and panic and fatal and exit in `main`. Let's just do the usual error handling here.

**Review by xh3b4sd — APPROVED** [review]
As we discussed, when the rest is addressed separately, this here is cool. Thanks a lot!

**Pull request merged**